### DISCUSSION TITLE
spark how to set instances

### DISCUSSION THREAD

**Discussion by A-little-bit-of-data** [apache/kyuubi#4596] [Q&A]
When I configure the following configuration in kyuubi-defaults.conf
spark.executor.instances=2
spark.driver.cores=1
spark.executor.cores=2
spark.driver.memory=1g
spark.executor.memory=2g

but the truth is 
![image](https://user-images.githubusercontent.com/31812278/227481371-e1caa806-e9f9-426a-af0b-66779437661c.png)

It seems that only spark.executor.instances is not taking effect，I connect via ./bin/beeline -u 'jdbc:hive2://localhost:10009/' -n kyuubi，Which partners have encountered similar problems, welcome to advise, thank you!

**Comment by zwangsheng** [comment]
Can you give much more info about `spark properties` from Spark UI Env Tab Or give `spark submit command` from Kyuubi Server Log?

Too little information not sure what the problem is

**Reply by A-little-bit-of-data** [reply]
Hello, I have not made any changes to the spark properties here, which are included when installing kyuubi; the log of kyuubi-server is as follows:
/opt/kyuubi/externals/spark-3.3.2-bin-hadoop3/bin/spark-submit \

  --class org.apache.kyuubi.engine.spark.SparkSQLEngine \

  --conf spark.hive.server2.thrift.resultset.default.fetch.size=1000 \

  --conf spark.kyuubi.client.ipAddress=XX.XX.XXX.XXX \

  --conf spark.kyuubi.client.version=1.7.0 \

  --conf spark.kyuubi.engine.share.level=USER \

  --conf spark.kyuubi.engine.submit.time=1679879887052 \

  --conf spark.kyuubi.frontend.min.worker.threads=500 \

  --conf spark.kyuubi.frontend.protocols=THRIFT_BINARY \

  --conf spark.kyuubi.ha.addresses=kyuubi-76b8676769-st95w:2181 \

  --conf spark.kyuubi.ha.engine.ref.id=833b0e2d-4f66-4150-91bf-7469999b64eb \

  --conf spark.kyuubi.ha.namespace=/kyuubi_1.7.0_USER_SPARK_SQL/trino1/default \

  --conf spark.kyuubi.ha.zookeeper.auth.type=NONE \

  --conf spark.kyuubi.kubernetes.namespace=XXXX\

  --conf spark.kyuubi.server.ipAddress=127.0.0.1 \

  --conf spark.kyuubi.session.connection.url=localhost:10009\

  --conf spark.kyuubi.session.engine.check.interval=PT3M \

  --conf spark.kyuubi.session.engine.idle.timeout=PT5M \

  --conf spark.kyuubi.session.real.user=trino1 \

  --conf spark.app.name=kyuubi_USER_SPARK_SQL_trino1_default_833b0e2d-4f66-4150-91bf-7469999b64eb \

  --conf spark.driver.cores=2 \

  --conf spark.driver.memory=1g \

  --conf spark.dynamicAllocation.enabled=false \

  --conf spark.executor.cores=2 \

  --conf spark.executor.instances=4 \

  --conf spark.executor.memory=4g \

  --conf spark.hadoop.fs.s3a.access.key=6NTWJGZLSK1Y9F38WC48 \

  --conf spark.hadoop.fs.s3a.connection.ssl.enabled=false \

  --conf spark.hadoop.fs.s3a.endpoint=http://XX.XX.X.XX:9000 \

  --conf spark.hadoop.fs.s3a.path.style.access=true \

  --conf spark.hadoop.fs.s3a.secret.key=XXX \

  --conf spark.hive.metastore.uris=thrift://XXXXX:9083 \

  --conf spark.kubernetes.driver.label.kyuubi-unique-tag=833b0e2d-4f66-4150-91bf-7469999b64eb \

  --conf spark.master=spark://XXXXX:7077 \

  --conf spark.yarn.tags=KYUUBI,833b0e2d-4f66-4150-91bf-7469999b64eb \

  --proxy-user trino1 /opt/kyuubi/externals/engines/spark/kyuubi-spark-sql-engine_2.12-1.7.0.jar

Then the corresponding webui is as follows
![image](https://user-images.githubusercontent.com/31812278/227819044-3c3f0d57-4dae-48fb-8c3a-bbd958b100c5.png)

The executors I applied for are only 4, and the information given in the log is indeed 4, but it does become 5 in the cluster. I really don't know why this is, thank you！

**Reply by zwangsheng** [reply]
> --conf spark.master=spark://XXXXX:7077 \

@A-little-bit-of-data 
Seems you are using `Standalone` cluster, I looked in the logic for assigning executors(in `org.apache.spark.deploy.master.Master#allocateWorkerResourceToExecutors`), find following code logic 
```
    // If the number of cores per executor is specified, we divide the cores assigned
    // to this worker evenly among the executors with no remainder.
    // Otherwise, we launch a single executor that grabs all the assignedCores on this worker.
    val numExecutors = coresPerExecutor.map { assignedCores / _ }.getOrElse(1)
```

It looks like it is not limited by the configuration you provide(spark.executor.instance), but rather the ratio of application-allocated resources to allocatable resources. Maybe there is a problem with my perception, you can check the corresponding code and thank you for the correction if there is an error.
